napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -0,0 +1,212 @@
+"""Plugin related functions that require Qt.
+
+Non-Qt plugin functions can be found in: `napari/plugins/_npe2.py`
+"""
+from __future__ import annotations
+
+import inspect
+from typing import (
+    TYPE_CHECKING,
+    List,
+    NamedTuple,
+    Optional,
+    Tuple,
+    Union,
+    cast,
+)
+
+from app_model import Action
+from app_model.types import SubmenuItem, ToggleRule
+from magicgui.type_map._magicgui import MagicFactory
+from magicgui.widgets import FunctionGui, Widget
+from npe2 import plugin_manager as pm
+from qtpy.QtWidgets import QWidget
+
+from napari._app_model import get_app
+from napari._app_model.constants import MenuGroup, MenuId
+from napari._qt._qapp_model.qactions._qproviders import _provide_window
+from napari.plugins import menu_item_template
+from napari.plugins._npe2 import (
+    _get_contrib_parent_menu,
+    get_widget_contribution,
+)
+from napari.utils.translations import trans
+from napari.viewer import Viewer
+
+if TYPE_CHECKING:
+    from npe2.manifest import PluginManifest
+    from npe2.plugin_manager import PluginName
+    from npe2.types import WidgetCreator
+
+    from napari._qt.qt_main_window import Window
+
+
+def _get_widget_viewer_param(
+    widget_callable: WidgetCreator, widget_name: str
+) -> str:
+    """Get widget parameter name (if any) and check type."""
+    if inspect.isclass(widget_callable) and issubclass(
+        widget_callable,
+        (QWidget, Widget),
+    ):
+        widget_param = ""
+        # Inspection can fail when adding to bundle as it thinks widget is
+        # a builtin
+        try:
+            sig = inspect.signature(widget_callable.__init__)
+        except ValueError:
+            pass
+        else:
+            for param in sig.parameters.values():
+                if param.name == 'napari_viewer' or param.annotation in (
+                    'napari.viewer.Viewer',
+                    Viewer,
+                ):
+                    widget_param = param.name
+                    break
+
+    elif isinstance(widget_callable, MagicFactory) or inspect.isfunction(
+        widget_callable
+    ):
+        widget_param = ""
```

Comment:
I think this merits a comment, that these widgets can request the viewer, but it's provided by magicgui(?) instead.

Reply:
Ah yes, I did previously have a comment but it somehow got lost again?

Reply:
Ah I have it in the docstring above. Do you think it should be here instead/as well?

Reply:
I've added a note here too